java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestNavigationUtils.java

```diff
@@ -0,0 +1,223 @@
+import org.eclipse.jdt.core.search.IJavaSearchScope;
+import org.eclipse.jdt.core.search.SearchEngine;
+import org.eclipse.jdt.core.search.SearchPattern;
+import org.eclipse.jdt.core.search.TypeNameRequestor;
+import org.eclipse.jdt.ls.core.internal.JDTUtils;
+import org.eclipse.jdt.ls.core.internal.JDTUtils.LocationType;
+import org.eclipse.jdt.ls.core.internal.ProjectUtils;
+import org.eclipse.lsp4j.Location;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Utils for test navigation features
+ */
+public class TestNavigationUtils {
+    /**
+     * find test or test target according to the given java source file uri.
+     * @param arguments arguments
+     * @param monitor monitor
+     * @return the search result for test navigation
+     * @throws JavaModelException
+     */
+    public static TestNavigationResult findTestOrTarget(List<Object> arguments, IProgressMonitor monitor)
+            throws JavaModelException {
+        if (arguments == null || arguments.size() < 2) {
+            throw new IllegalArgumentException("Wrong arguments passed to findTestOrTarget().");
+        }
+        final String typeUri = (String) arguments.get(0);
+        final ICompilationUnit unit = JDTUtils.resolveCompilationUnit(typeUri);
+        if (unit == null) {
+            JUnitPlugin.logError("Failed to resolve compilation unit from " + typeUri);
+            return null;
+        }
+        final boolean goToTest = (boolean) arguments.get(1);
+        final IType primaryType = unit.findPrimaryType();
+        final String typeName;
+        Location location = null;
+        if (primaryType != null) {
+            typeName = primaryType.getElementName();
+            location = JDTUtils.toLocation(primaryType, LocationType.NAME_RANGE);
+        } else {
+            typeName = unit.getElementName().substring(0, unit.getElementName().lastIndexOf(".java"));
+        }
+        final SearchEngine searchEngine = new SearchEngine();
+        final IJavaProject javaProject = unit.getJavaProject();
+        final IJavaSearchScope scope = goToTest ? getSearchScopeForTest() :
+                getSearchScopeForTarget();
+        final Set<TestNavigationItem> items = new HashSet<>();
+        searchEngine.searchAllTypeNames(
+            null,
+            SearchPattern.R_EXACT_MATCH,
+            ("*" + typeName + "*").toCharArray(),
+            SearchPattern.R_PREFIX_MATCH,
+            IJavaSearchConstants.CLASS,
+            scope,
+            new TestNavigationNameRequestor(items, javaProject, typeName),
+            IJavaSearchConstants.WAIT_UNTIL_READY_TO_SEARCH,
+            monitor
+        );
+
+        return new TestNavigationResult(items, location);
+    }
+
+    private static IJavaSearchScope getSearchScopeForTarget() {
+        // TODO: unimplemented
```

Comment:
won't be implemented in this PR? any issue to track?

Reply:
There will be a separate PR. Tracked by https://github.com/microsoft/vscode-java-test/issues/660